lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1712,137 @@ def find(root, files, recursive=True):
+            # example if we are scanning an external directory like /usr
+            continue
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+        with dir_iter:
+            for dir_entry in dir_iter:
+                orig_path = os.path.join(next_dir, dir_entry.name)
+
+                try:
+                    it_is_a_dir = dir_entry.is_dir(follow_symlinks=True)
+                except OSError as e:
+                    if e.errno == errno.ELOOP:
+                        tty.debug(
+                            f"{dir_entry.path}: this path is a symlink that "
+                            "cannot be resolved - skipping"
+                        )
+                        continue
+                    else:
+                        raise
+
+                if it_is_a_dir:
+                    resolved_path = None
+                    if sys.platform == "win32":
+                        # Note: DirEntry.is_junction is available starting with python 3.12
+                        # but this must work for earlier versions
+                        if islink(dir_entry.path):
+                            resolved_path = os.path.realpath(readlink(dir_entry.path))
+                    else:
+                        if dir_entry.is_symlink():
+                            # Note: we want the resolved path to avoid more symlink
+                            # resolutions when running `scandir` on children of this dir.
+                            # We already followed the symlinks once to determine
+                            # if this is a directory, so if we're here, it would have
+                            # been more efficient to call os.realpath and then check
+                            # if the result was a dir. However, if it *had been* a file,
+                            # then calling `isdir` would have been extra work.
+                            resolved_path = os.path.realpath(dir_entry.path)
+
+                    if not resolved_path:
+                        # If resolved_path hasn't been assigned yet, it's not a link
+                        resolved_path = os.path.join(next_dir_resolved, dir_entry.name)
+
+                    if len(resolved_path) < len(next_dir_resolved):
+                        # Either the symlink target is entirely outside the root,
+                        # or it will be accounted for "directly"
+                        continue
+
+                    if (depth < max_depth) and (resolved_path not in visited_dirs):
+                        dir_queue.appendleft((depth + 1, orig_path, resolved_path))
+                        visited_dirs.add(resolved_path)
+                else:
+                    fname = os.path.basename(orig_path)
+                    for pattern in regexes:
+                        if pattern.match(os.path.normcase(fname)):
+                            found_files[pattern].append(os.path.join(next_dir, fname))
```

Comment:
Can you replace `for pattern in regexes: pattern.match(...)` with a single regex using named capture groups? We use that everywhere else.

Reply:
IMO that pattern should be collected into a library function (that we unit test) and then that function should be used here. There's a few things to get wrong with this type of regex squashing and I don't want to introduce those complexities into this PR.